src/irmin/tree.ml

```diff
@@ -511,61 +502,63 @@ module Make (P : Private.S) = struct
         let v =
           StepMap.to_seq map
           |> Seq.map (function
-               | step, `Contents (c, m) -> (step, `Contents (Contents.hash c, m))
-               | step, `Node n -> (step, hash n (fun h -> `Node h)))
+               | step, `Contents (c, m) ->
+                   (step, `Contents (Contents.hash ~cache c, m))
+               | step, `Node n -> (step, hash ~cache n (fun h -> `Node h)))
           |> P.Node.Val.of_seq
         in
-        t.info.value <- Some v;
+        if cache then t.info.value <- Some v;
         k v)
 
-    and value_of_elt : type r. elt -> (P.Node.Val.value, r) cont =
-     fun e k ->
+    and value_of_elt : type r. cache:bool -> elt -> (P.Node.Val.value, r) cont =
+     fun ~cache e k ->
       match e with
-      | `Contents (c, m) -> k (`Contents (Contents.hash c, m))
-      | `Node n -> hash n (fun h -> k (`Node h))
+      | `Contents (c, m) -> k (`Contents (Contents.hash ~cache c, m))
+      | `Node n -> hash ~cache n (fun h -> k (`Node h))
 
-    and value_of_updates : type r. t -> value -> _ -> (value, r) cont =
-     fun t v updates k ->
+    and value_of_updates :
+        type r. cache:bool -> t -> value -> _ -> (value, r) cont =
+     fun ~cache t v updates k ->
       let updates = StepMap.bindings updates in
       let rec aux acc = function
         | [] ->
-            t.info.value <- Some acc;
+            if cache then t.info.value <- Some acc;
             k acc
         | (k, Add e) :: rest ->
-            value_of_elt e (fun e -> aux (P.Node.Val.add acc k e) rest)
+            value_of_elt ~cache e (fun e -> aux (P.Node.Val.add acc k e) rest)
         | (k, Remove) :: rest -> aux (P.Node.Val.remove acc k) rest
       in
       aux v updates
 
-    let hash k = hash k (fun x -> x)
+    let hash ~cache k = hash ~cache k (fun x -> x)
 
-    let value_of_hash t repo k =
+    let value_of_hash ~cache t repo k =
       match cached_value t with
       | Some v -> Lwt.return_ok v
       | None -> (
           cnt.node_find <- cnt.node_find + 1;
           P.Node.find (P.Repo.node_t repo) k >|= function
           | None -> Error (`Dangling_hash k)
           | Some v as some_v ->
-              t.info.value <- some_v;
+              if cache then t.info.value <- some_v;
               Ok v)
 
-    let to_value t =
+    let to_value ~cache t =
       match cached_value t with
       | Some v -> ok v
       | None -> (
           match t.v with
           | Value (_, v, None) -> ok v
-          | Value (_, v, Some um) -> value_of_updates t v um ok
-          | Map m -> value_of_map t m ok
-          | Hash (repo, h) -> value_of_hash t repo h)
+          | Value (_, v, Some um) -> value_of_updates ~cache t v um ok
+          | Map m -> value_of_map ~cache t m ok
+          | Hash (repo, h) -> value_of_hash ~cache t repo h)
 
-    let to_map t =
+    let to_map ~cache t =
```

Comment:
here `cache` is used to control both the caching of the generated map and of any value which might be read. Do we want to distinguish both? (I'm not sure it's useful, just wondering)

Reply:
I suppose we could change all the `~cache` parameter to a flag that controls each of the cache entries:
- `info.hash`
- `info.value`
- `info.findv_cache`
- `info.map`
- node backend

I don't see a use case now

Reply:
> I don't see a use case now